_posts/2017-09-18-rnn-bucket-mxnet-R.md

```diff
@@ -0,0 +1,206 @@
+
+optimizer <- mx.opt.create("rmsprop", learning.rate = 1e-3, gamma1 = 0.95, gamma2 = 0.95, 
+                           wd = 1e-4, clip_gradient = 5, rescale.grad=1/batch.size)
+
+logger <- mx.metric.logger()
+epoch.end.callback <- mx.callback.log.train.metric(period = 1, logger = logger)
+batch.end.callback <- mx.callback.log.train.metric(period = 50)
+
+system.time(
+  model <- mx.model.buckets(symbol = symbol_single,
+                            train.data = train.data.single, eval.data = eval.data.single,
+                            num.round = 5, ctx = devices, verbose = FALSE,
+                            metric = mx.metric.accuracy, optimizer = optimizer,  
+                            initializer = initializer,
+                            batch.end.callback = NULL, 
+                            epoch.end.callback = epoch.end.callback)
+)
+```
+
+    ##    user  system elapsed 
+    ## 161.372  20.650 172.173
+
+![](https://raw.githubusercontent.com/dmlc/web-data/master/mxnet/blog_mxnet_R_rnn_bucket/logger1-1.png)
+
+Then training with the bucketing trick. Note that no additional effort is required: just need to provide a list of symbols rather than a single one and have an iterator pushing samples from the different buckets.
+
+``` r
+devices <- mx.gpu(0)
+
+initializer <- mx.init.Xavier(rnd_type = "gaussian", factor_type = "avg", magnitude = 2.5)
+
+optimizer <- mx.opt.create("rmsprop", learning.rate = 1e-3, gamma1 = 0.95, gamma2 = 0.95, 
+                           wd = 1e-4, clip_gradient = 5, rescale.grad=1/batch.size)
+
+logger <- mx.metric.logger()
+epoch.end.callback <- mx.callback.log.train.metric(period = 1, logger = logger)
+batch.end.callback <- mx.callback.log.train.metric(period = 50)
+
+system.time(
+  model <- mx.model.buckets(symbol = symbol_buckets,
+                            train.data = train.data.bucket, eval.data = eval.data.bucket,
+                            num.round = 5, ctx = devices, verbose = FALSE,
+                            metric = mx.metric.accuracy, optimizer = optimizer,  
+                            initializer = initializer,
+                            batch.end.callback = NULL, 
+                            epoch.end.callback = epoch.end.callback)
+)
+```
+
+    ##    user  system elapsed 
+    ##  96.690  15.855  99.760
+
+``` r
+mx.model.save(model, prefix = "models/model_sentiment_lstm", iteration = 5)
+```
+
+![](https://raw.githubusercontent.com/dmlc/web-data/master/mxnet/blog_mxnet_R_rnn_bucket/logger2-1.png)
+
+The speedup is substantial, around 100 sec. instead of 175 sec., a 40% speedup with little effort.
+
+Plot word embeddings
+--------------------
+
+Word representation can be visualized by looking at the assigned weights in any of the embedding dimensions. Here, we look simultaneously at the two embeddings learnt in the LSTM model.
+
+![](https://raw.githubusercontent.com/dmlc/web-data/master/mxnet/blog_mxnet_R_rnn_bucket/embed-1.png)
```

Comment:
1. For "1" and "3" in the plot, it might be more interesting to list some other numbers as well, or compare them with the words "one" and "three".
2. It is good to have one or two sentences mentioning some interesting observations (e.g. the "excellent" and "perfect" are close) in the plot.